modules/camera/camera_x11.cpp

```diff
@@ -0,0 +1,770 @@
+
+			r = select(fd + 1, &fds, NULL, NULL, &tv);
+		} while (-1 == r && (EINTR == errno || EAGAIN == errno));
+		if (r <= 0) {
+			return;
+		}
+
+		// grab image depending on the type of image grabbing
+		// currently only tested for TYPE_IO_MMAP
+		switch (type) {
+			case TYPE_IO_READ: {
+				if (-1 == funcs->read(fd, buffers[0].start, buffers[0].length)) {
+					switch (errno) {
+						case EAGAIN:
+							continue;
+						case EIO: // could be ignored somehow?
+						default:
+							return;
+					}
+				}
+
+				get_image(feed, (uint8_t *)(buffers[0].start));
+				break;
+			}
+			case TYPE_IO_MMAP: {
+				CLEAR(buf);
+
+				buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+				buf.memory = V4L2_MEMORY_MMAP;
+
+				if (-1 == xioctl(fd, VIDIOC_DQBUF, &buf)) {
+					switch (errno) {
+						case EAGAIN:
+							continue;
+						case EIO: // could be ignored
+						default:
+							return;
+					}
+				}
+
+				get_image(feed, (uint8_t *)(buffers[buf.index].start));
+
+				if (-1 == xioctl(fd, VIDIOC_QBUF, &buf))
+					return;
+				break;
+			}
+			case TYPE_IO_USRPTR: {
+				CLEAR(buf);
+
+				buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+				buf.memory = V4L2_MEMORY_USERPTR;
+
+				if (-1 == xioctl(fd, VIDIOC_DQBUF, &buf)) {
+					switch (errno) {
+						case EAGAIN:
+							continue;
+						case EIO: // could be ignored
+						default:
+							return;
+					}
+				}
+
+				for (unsigned int i = 0; i < n_buffers; ++i)
+					if (buf.m.userptr == (unsigned long)buffers[i].start && buf.length == buffers[i].length)
+						break;
+
+				get_image(feed, (uint8_t *)(buf.m.userptr));
+
+				if (-1 == xioctl(fd, VIDIOC_QBUF, &buf))
+					return;
+				break;
+			}
+			default:
+				break;
+		}
+	}
+}
+
+void V4l2_Device::get_image(Ref<CameraFeed> feed, uint8_t *buffer) {
+	Ref<Image> img;
+	img.instance();
+
+	if (-1 == xioctl(fd, VIDIOC_G_FMT, &fmt)) {
+		return;
+	}
+
+	// other format implementations should be put here.
+	// Mainly useful if libv4l2 is not installed
+	switch (fmt.fmt.pix.pixelformat) {
+		case V4L2_PIX_FMT_RGB24: {
+			if (width != fmt.fmt.pix.width || height != fmt.fmt.pix.height) {
+				width = fmt.fmt.pix.width;
+				height = fmt.fmt.pix.height;
+				img_data.resize(width * height * 3);
+			}
+
+			uint8_t *w = img_data.ptrw();
+			// TODO: Buffer is 1024 Byte longer?
+			memcpy(w, buffer, width * height * 3);
+
+			img->create(width, height, 0, Image::FORMAT_RGB8, img_data);
+			feed->set_RGB_img(img);
+			break;
+		}
+		default:
+			break;
+	}
+}
+
+//////////////////////////////////////////////////////////////////////////
+// CameraFeedX11 - Subclass for camera feeds in Linux
+
+V4l2_Device *CameraFeedX11::get_device() const {
+	return device;
+};
+
+CameraFeedX11::CameraFeedX11() {
+	device = NULL;
+};
+
+void CameraFeedX11::set_device(V4l2_Device *p_device) {
+	device = p_device;
+
+	// get some info
+	name = device->name;
+	// I do not think that another position is possible on linux
+	position = CameraFeed::FEED_UNSPECIFIED;
```

Comment:
Unspecified I think is fine for a webcam, front/back really only applies to phones.